modules/xfeatures2d/src/msd.cpp

```diff
@@ -0,0 +1,791 @@
+                std::vector<float *> saliency;
+                saliency.resize(m_cur_n_scales);
+
+                for (int r = 0; r < m_cur_n_scales; r++)
+                {
+                    saliency[r] = new float[m_scaleSpace[r].rows * m_scaleSpace[r].cols];
+                }
+
+                for (int r = 0; r < m_cur_n_scales; r++)
+                {
+                    int steps = cv::getNumThreads();
+                    parallel_for_(Range(0, steps), MSDSelfDissimilarityScan((*this), &saliency, m_scaleSpace[r], r, border, steps));
+                }
+
+                nonMaximaSuppression(saliency, keypoints);
+
+                for (int r = 0; r < m_n_scales; r++)
+                {
+                    delete[] saliency[r];
+                }
+
+                m_scaleSpace.clear();
+
+            }
+
+        private:
+
+
+
+            std::vector<cv::Mat> m_scaleSpace;
+            cv::Mat m_mask;
+
+            inline float computeAvgDistance(std::vector<int> &minVals, int den)
+            {
+                float avg_dist = 0.0f;
+                for (unsigned int i = 0; i < minVals.size(); i++)
+                    avg_dist += minVals[i];
+
+                avg_dist /= den;
+                return avg_dist;
+            }
+
+            void contextualSelfDissimilarity(cv::Mat &img, int xmin, int xmax, float* saliency);
+
+            float computeOrientation(cv::Mat &img, int x, int y, std::vector<cv::Point2f> circle);
+
+            void nonMaximaSuppression(std::vector<float *> & saliency, std::vector<cv::KeyPoint> & keypoints);
+
+            bool rescalePoint(int x, int y, int scale, std::vector<float *> & saliency, cv::Point2f & p_res);
+
+        };
+
+        bool MSDDetector_Impl::rescalePoint(int i, int j, int scale, std::vector<float *> & saliency, cv::Point2f &p_res)
+        {
+
+            const float deriv_scale = 0.5f;
+            int width_s = m_scaleSpace[scale].cols;
+            //const float second_deriv_scale = 1.0f;
+            const float cross_deriv_scale = 0.25f;
+
+            cv::Vec2f dD((saliency[scale][j * width_s + i + 1] - saliency[scale][j * width_s + i - 1]) * deriv_scale,
+                    (saliency[scale][(j + 1) * width_s + i] - saliency[scale][(j - 1) * width_s + i]) * deriv_scale);
+
+            float cc = saliency[scale][j * width_s + i] * 2;
+            float dxx = (saliency[scale][j * width_s + i + 1] + saliency[scale][j * width_s + i - 1] - cc);// * second_deriv_scale;
+            float dyy = (saliency[scale][(j + 1) * width_s + i] + saliency[scale][(j - 1) * width_s + i] - cc);// * second_deriv_scale;
+            float dxy = (saliency[scale][(j + 1) * width_s + i + 1] - saliency[scale][(j + 1) * width_s + i - 1] -
+                    saliency[scale][(j - 1) * width_s + i + 1] + saliency[scale][(j - 1) * width_s + i - 1]) * cross_deriv_scale;
+            
+            cv::Matx22f H(dxx, dxy, dxy, dyy);
+
+            cv::Vec2f X = H.solve(dD, cv::DECOMP_LU);
```

Comment:
My guess is that the solve operator returns a double precision result, while you are trying to capture it in a floating point precision object. Could you try changing the type of X to be double precision?